pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,267 @@
+    """Reads out the generator and prefactor of an operation and converts
+    to matrix if necessary.
+
+    Args:
+        op (:class:`~.Operation`): Operation to obtain the generator of.
+    Returns:
+        array[float]: Generator matrix
+        float: Prefactor of the generator
+    """
+
+    generator, prefactor = op.generator
+    if not isinstance(generator, np.ndarray):
+        generator = generator.matrix
+    if op.inverse:
+        generator = generator.conj().T
+        prefactor *= -1
+
+    return generator, prefactor
+
+
+def _apply_any_operation(state, op, device, invert=False):
+    """Wrapper that allows to apply a variety of operations---or groups
+    of operations---to a state or to prepare a new state.
+    If ``invert=True``, this function makes sure not to alter the operations.
+    The state of the device, however may be altered, depending on the
+    device and performed operation(s).
+    """
+    # pylint: disable=protected-access
+    if isinstance(op, (list, np.ndarray)):
+        if invert:
+            op = op[::-1]
+        for _op in op:
+            state = _apply_any_operation(state, _op, device, invert)
+        return state
+    if isinstance(op, qml.QubitStateVector):
+        if invert:
+            raise ValueError("Can't invert state preparation.")
+        device._apply_state_vector(op.parameters[0], op.wires)
+        return device._state
+    if isinstance(op, qml.BasisState):
+        if invert:
+            raise ValueError("Can't invert state preparation.")
+        device._apply_basis_state(op.parameters[0], op.wires)
+        return device._state
+
+    if invert:
+        op_was_inverted = op.inverse
+        op = qml.adjoint(op.__class__)(*op.data, wires=op.wires)
+        if op_was_inverted:
+            op.inverse = not op.inverse
+    state = device._apply_operation(state, op)
+
+    return state
+
+
+def _group_operations(tape):
+    """Divide all operations of a tape into trainable operations and blocks
+    of untrainable operations after each trainable one."""
+
+    # Extract tape operations list
+    ops = tape.operations
+    # Find the indices of trainable operations in the tape operations list
+    trainables = np.where([qml.operation.is_trainable(op) for op in ops])[0]
+    # Add the indices incremented by one to the trainable indices
+    split_ids = list(chain.from_iterable([idx, idx + 1] for idx in trainables))
+
+    # Split at trainable and incremented indices to get groups after trainable
+    # operations and single trainable operations (in alternating order)
+    all_groups = np.split(ops, split_ids)
+
+    # Collect trainable operations and groups after trainable operations
+    # the first set of non-trainable ops are the ops "after the -1st" trainable op
+    group_after_trainable_op = dict(enumerate(all_groups[::2], start=-1))
+    trainable_operations = list(chain.from_iterable(all_groups[1::2]))
+
+    return trainable_operations, group_after_trainable_op
```

Comment:
I see how this is confusing. 
1. Think of these arrays as lists. They just are not lists because they are output of `np.split`.
2. The values in the dictionary are groups of _non_trainable operations. The key gives the index of the trainable operation after which the respective group of nontrainable ones follows. In particular, the very first group follows the - nonexistent - `-1`st trainable operation, i.e it contains those non-trainable operations that appear before the `0`th trainable one. Going through your example (calling the dictionary `d`):
-1 -> There is a non-trainable `RX` gate in the beginning, it is contained in `d[-1]`.
0 -> There are no non-trainable operations that come after the `0`th trainable operation but before the `1`st trainable one. Therefore `d[0]` is empty.
1 -> After the `1`st and before the `2`nd trainable operation there is a `CNOT`, which is in `d[1]`.
2 -> Similar to `d[1]`, but the non-trainable operation is not a `CNOT` but an `RX`.
3 -> After the `3`rd - and last - trainable operation, there is a final non-trainable `CNOT`, which is contained in `d[3]`.

Reply:
thanks for the explanation, makes a lot of sense!